### PR TITLE
Adding data2vec to atari decision transformers

### PR DESCRIPTION
**Pull request by vincentpierre**
running `run_dt_atari.py` with the flag `--data2vec` will add a [data2vec](https://ai.facebook.com/blog/the-first-high-performance-self-supervised-algorithm-that-works-for-speech-vision-and-text/) loss term to the decision transformer loss during training.

The current masking for the student used is masking all the actions. There are other (better) ways to mask the student input, like masking a sub-sequence of observations, actions and rtg.

Limitations : 
 - data2vec is supposed to be a pretraining step in supervised learning methods. Here the loss terms of DT and data2vec are added together (for experimentation).

### CODE REVIEW

File: atari/mingpt/model_atari.py [hunk 1]
```diff
@@ -111,11 +114,90 @@ def __init__(self, config):
...
         )
 
-    def forward(self, x):
-        x = x + self.attn(self.ln1(x))
+    def forward(self, x, causal = True):
+        x = x + self.attn(self.ln1(x), causal=causal)
         x = x + self.mlp(self.ln2(x))
         return x
 
+def soft_update(target: torch.nn.Module, source: torch.nn.Module, tau: float) -> None:
```
Comment (vincentpierre) [inline atari/mingpt/model_atari.py:122]: needed for updating the teacher network with the student weights.

File: atari/mingpt/model_atari.py [hunk 2]
```diff
@@ -111,11 +114,90 @@ def __init__(self, config):
...
+        """
+        Performs an in-place polyak update of the target module based on the source,
+        by a ratio of tau. Note that source and target modules must have the same
+        parameters, where:
+            target = tau * source + (1-tau) * target
```
Comment (vincentpierre) [inline atari/mingpt/model_atari.py:127]:
```suggestion
            target = (1-tau) * source + tau * target
```

File: atari/mingpt/model_atari.py [hunk 3]
```diff
@@ -111,11 +114,90 @@ def __init__(self, config):
...
+                target_param.data.mul_(tau)
+                torch.add(
+                    target_param.data,
+                    source_param.data,
+                    alpha=1-tau,
+                    out=target_param.data,
+                )
+
+
+def data_to_vec_loss(student, teacher, token_embeddings, position_embeddings, mask_embedding, k, tau = 0.9998, beta=2.0):
```
Comment (vincentpierre) [inline atari/mingpt/model_atari.py:146]: The core of the code. This should be the same loss as described in data2vec

File: atari/mingpt/trainer_atari.py [hunk 4]
```diff
@@ -100,15 +104,15 @@ def run_epoch(split, epoch_num=0):
...
+                    logits, loss, data2vec_loss = model(x, y, y, r, t)
                     loss = loss.mean() # collapse all losses if they are scattered on multiple gpus
                     losses.append(loss.item())
 
                 if is_train:
 
                     # backprop and update the parameters
                     model.zero_grad()
-                    loss.backward()
+                    (loss + data2vec_loss.mean()).backward()
```
Comment (vincentpierre) [inline atari/mingpt/trainer_atari.py:115]: adding the DT loss with the data2vec loss for now.

File: atari/my_run_dt_atari.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import csv
```
Comment (vincentpierre) [inline atari/my_run_dt_atari.py:1]: this is code to launch a batch of experiments with a python script.
